src/equisolve/utils/split_data.py

```diff
@@ -0,0 +1,431 @@
+    != -1, shuffling is executed but with a random seed set to this value. There
+    is no return value, as the input `indices` is shuffled in place.
+    """
+    if random_seed is not None:  # shuffle
+        if random_seed != -1:  # set a numpy random seed
+            np.random.seed(random_seed)
+        np.random.shuffle(indices)
+
+
+def _get_group_sizes(
+    n_groups: int,
+    n_indices: int,
+    group_sizes: Optional[Union[List[float], List[int]]] = None,
+) -> np.ndarray:
+    """
+    Parses the `group_sizes` arg from :py:func:`split_data` and returns an array
+    of group sizes in absolute numbers.
+
+    :param n_groups: an int, the number of groups to split the data into :param
+        n_indices: an int, the number of unique indices present in the data by
+        which the data should be grouped.
+    :param n_indices: a :py:class:`int` for the number of unique indices present
+        in the input data for the specified `axis` and `names`.
+    :param group_sizes: a :py:class:`list` of :py:class:`float` or
+        :py:class:`int` indicating the absolute or relative group sizes,
+        respectively.
+
+    :return: a :py:class:`numpy.ndarray` of :py:class:`int` indicating the
+        absolute group sizes.
+    """
+    if group_sizes is None:  # equally sized groups
+        group_sizes = np.array([1 / n_groups] * n_groups) * n_indices
+    elif np.all([isinstance(size, int) for size in group_sizes]):  # absolute
+        group_sizes = np.array(group_sizes)
+    else:  # relative; list of float
+        group_sizes = np.array(group_sizes) * n_indices
+
+    # The group sizes may not be integers. Use cascade rounding to round them
+    # all to integers whilst attempting to minimize rounding error.
+    group_sizes = _cascade_round(group_sizes)
+
+    return group_sizes
+
+
+def _cascade_round(array: np.ndarray) -> np.ndarray:
+    """
+    Given an array of floats that sum to an integer, this rounds the floats
+    and returns an array of integers with the same sum.
+    Adapted from https://jsfiddle.net/cd8xqy6e/.
+    """
+    # Check type
+    if not isinstance(array, np.ndarray):
+        raise TypeError("must pass `array` as a numpy array.")
+    # Check sum
+    mod = np.sum(array) % 1
+    if not np.isclose(round(mod) - mod, 0):
+        raise ValueError("elements of `array` must sum to an integer.")
+
+    float_tot, integer_tot = 0, 0
+    rounded_array = []
+    for element in array:
+        new_int = round(element + float_tot) - integer_tot
+        float_tot += element
+        integer_tot += new_int
+        rounded_array.append(new_int)
+
+    # Check that the sum is preserved
+    assert round(np.sum(array)) == round(np.sum(rounded_array))
+
+    return np.array(rounded_array)
+
+
+def _group_indices(indices: Labels, group_sizes: List[int]) -> Labels:
+    """
+    Sets a numpy `random_seed` then randomly shuffles `indices`. Next, these
```

Comment:
There is nothing shuffled in this function or?

Reply:
Good spot!